server/server.go

```diff
@@ -977,6 +1005,126 @@ func (s *Server) etcdLeaderLoop() {
 	}
 }
 
+func (s *Server) configCheckLoop() {
+	defer logutil.LogPanic()
+	defer s.serverLoopWg.Done()
+
+	ctx, cancel := context.WithCancel(s.serverLoopCtx)
+	defer cancel()
+
+	ticker := time.NewTicker(configCheckInterval)
+	defer ticker.Stop()
+	for {
+		// wait leader
+		leader := s.GetLeader()
+		if leader != nil {
+			var err error
+			securityConfig := s.GetSecurityConfig()
+			s.configClient, err = pd.NewConfigClientWithContext(ctx, s.GetEndpoints(), pd.SecurityOption{
+				CAPath:   securityConfig.CAPath,
+				CertPath: securityConfig.CertPath,
+				KeyPath:  securityConfig.KeyPath,
+			})
+			if err != nil {
+				log.Error("failed to create config client", zap.Error(err))
+				return
+			}
+			break
+		}
+		select {
+		case <-ticker.C:
+		case <-ctx.Done():
+			log.Info("config check stop running")
+			return
+		}
+	}
+
+	addr := s.GetAddr()
+	version := s.GetConfigVersion()
+	var config bytes.Buffer
+	if err := toml.NewEncoder(&config).Encode(*s.GetConfig()); err != nil {
+		log.Error("failed to encode config", zap.Error(err))
+		return
+	}
+	if err := s.createComponentConfig(ctx, version, addr, config.String()); err != nil {
+		log.Error("failed to create config", zap.Error(err))
+		return
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("config check has been stopped")
+			return
+		case <-ticker.C:
+			version := s.GetConfigVersion()
+			config, err := s.getComponentConfig(ctx, version, addr)
+			if err != nil {
+				log.Error("failed to get config", zap.Error(err))
+			}
+			if config == "" {
+				continue
+			}
+			if err := s.updateComponentConfig(config); err != nil {
+				log.Error("failed to update config", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (s *Server) createComponentConfig(ctx context.Context, version *configpb.Version, componentID, config string) error {
+	status, v, config, err := s.configClient.Create(ctx, version, Component, componentID, config)
+	if err != nil {
+		return err
+	}
+	switch status.GetCode() {
+	case configpb.StatusCode_OK, configpb.StatusCode_WRONG_VERSION:
+		s.setConfigVersion(v)
+		s.updateComponentConfig(config)
+	case configpb.StatusCode_UNKNOWN:
+		return errors.Errorf("unknown error: %v", status.GetMessage())
+	}
+	return nil
+}
+
+func (s *Server) getComponentConfig(ctx context.Context, version *configpb.Version, componentID string) (string, error) {
+	status, v, cfg, err := s.configClient.Get(ctx, version, Component, componentID)
+	if err != nil {
+		return "", err
+	}
+	var config string
+	switch status.GetCode() {
+	case configpb.StatusCode_OK:
+		config = cfg
+	case configpb.StatusCode_WRONG_VERSION:
+		s.setConfigVersion(v)
```

Comment:
We should move L1098 to L1101, because when `StatusCode_OK` is return means there is no update, and `StatusCode_WRONG_VERSION` means the opposite.